src/compiler/java_compiler.nit

```diff
@@ -0,0 +1,2297 @@
+		if toolcontext.verbose_level >= 3 then
+			res = sys.system("make -B -C {compile_dir} -f {outname}.mk 2>&1")
+		else
+			res = sys.system("make -B -C {compile_dir} -f {outname}.mk 2>&1 > /dev/null")
+		end
+		if res != 0 then toolcontext.error(null, "make failed! Error code: {res}.")
+	end
+
+	# Compile Java sources using `ant`
+	fun build_with_ant(compiler: JavaCompiler, jfiles: Array[String]) do
+		compile_antfile(compiler, jfiles)
+		var outname = compiler.outname.to_path.filename
+		var antpath = "{compiler.compile_dir}/{outname}.xml"
+		self.toolcontext.info("ant jar -f {antpath}", 2)
+		var res
+		if self.toolcontext.verbose_level >= 3 then
+			res = sys.system("ant jar -f {antpath} 2>&1")
+		else
+			res = sys.system("ant jar -f {antpath} 2>&1 > /dev/null")
+		end
+		if res != 0 then
+			toolcontext.error(null, "ant compile failed! Error code: {res}.")
+		end
+	end
+
+	# Write the Makefile used to compile Java generated files into an executable jar
+	fun write_makefile(compiler: JavaCompiler, jfiles: Array[String]) do
+		# list class files from jfiles
+		var ofiles = new List[String]
+		for f in jfiles do ofiles.add(f.strip_extension(".java") + ".class")
+
+		var compile_dir = compiler.compile_dir
+		var outname = compiler.outname.to_path.filename
+		var outpath = (sys.getcwd / compiler.outname).simplify_path
+		var makename = "{compile_dir}/{outname}.mk"
+		var makefile = new FileWriter.open(makename)
+
+		makefile.write("JC = javac\n")
+		makefile.write("JAR = jar\n\n")
+
+		makefile.write("all: {outpath}.jar\n\n")
+
+		makefile.write("{outpath}.jar: {compiler.mainmodule.jname}_Main.class\n")
+		makefile.write("\t$(JAR) cfm {outpath}.jar {outname}.mf {ofiles.join(" ")}\n\n")
+
+		makefile.write("{compiler.mainmodule.jname}_Main.class:\n")
+		makefile.write("\t$(JC) {jfiles.join(" ")}\n\n")
+
+		makefile.write("clean:\n")
+		makefile.write("\trm {ofiles.join(" ")} 2>/dev/null\n\n")
+
+		makefile.close
+		toolcontext.info("Generated makefile: {makename}", 2)
+	end
+
+	# The Ant `build.xml` script used to compile build the final jar
+	fun compile_antfile(compiler: JavaCompiler, jfiles: Array[String]) do
+		var compile_dir = compiler.compile_dir
+		var outname = compiler.outname.to_path.filename
+		var outpath = (sys.getcwd / compiler.outname).simplify_path
+		var antname = "{compile_dir}/{outname}.xml"
+		var antfile = new FileWriter.open(antname)
+		var jname = compiler.mainmodule.jname
+		antfile.write("<project>")
+		antfile.write(" <target name=\"compile\">")
+		antfile.write("  <mkdir dir=\"classes\"/>")
+        antfile.write("  <javac includes=\"{compiler.mainmodule.jname}_Main.java {jfiles.join(" ")}\" srcdir=\".\" destdir=\"classes\"/>")
+		antfile.write(" </target>")
+		antfile.write(" <target name=\"jar\" depends=\"compile\">")
+		antfile.write("  <jar destfile=\"{outpath}.jar\" basedir=\"classes\">")
+		antfile.write("   <manifest>")
+        antfile.write("    <attribute name=\"Main-Class\" value=\"{jname}_Main\"/>")
```

Comment:
Weird indentation. IMO, this block would be cleaner in a multiline string or without parentheses to `write`.